arangod/Aql/Optimizer/Rule/OptimizerRuleCollectWithIndex.cpp

```diff
@@ -171,33 +215,139 @@ isEligiblePair(ExecutionPlan const& plan, CollectNode const& cn,
+      LOG_RULE << "Calculation " << calculation->id()
+               << " not eligible - expression uses other variables than index ("
+               << idx.id() << ") out variable " << idx.outVariable()->name;
+      return std::nullopt;
+    }
+
+    // check if all attributes in aggregate expression are covered fields
+    containers::FlatHashSet<aql::AttributeNamePath> attributes;
+    Ast::getReferencedAttributesRecursive(
+        calculation->expression()->node(), idx.outVariable(), "", attributes,
+        plan.getAst()->query().resourceMonitor());
+    for (auto const& attr : attributes) {
+      auto iter = std::find_if(
+          coveredFields.begin(), coveredFields.end(),
+          [&](auto const& field) { return attributeMatches(attr, field); });
+      if (iter == coveredFields.end()) {
+        LOG_RULE << "Calculation " << calculation->id()
+                 << " not eligible - accessed attribute " << attr
+                 << " which is not covered by the index";
+        return std::nullopt;
+      }
+      std::size_t fieldIndex = std::distance(coveredFields.begin(), iter);
+      aggregationFields.emplace_back(fieldIndex);
+    }
+
+    // put the calculation expression inside the aggregation
+    aggregations.emplace_back(IndexCollectAggregation{
+        .type = agg.type,
+        .outVariable = agg.outVar,
+        .expression = calculation->expression()->clone(plan.getAst())});
+  }
+
+  return std::make_tuple(std::move(aggregations), std::move(aggregationFields));
+}
+
+bool isIndexStreamSupported(IndexNode const& idx,
+                            std::vector<size_t> groupFields,
+                            std::vector<size_t> aggregationFields) {
+  IndexStreamOptions streamOptions{.usedKeyFields = std::move(groupFields),
+                                   .projectedFields = aggregationFields};
+  if (auto support =
+          idx.getSingleIndex()->supportsStreamInterface(streamOptions);
+      support.hasSupport()) {
+    return true;
+  }
+  LOG_RULE
+      << "Index " << idx.id()
+      << " not eligible - index does not support required stream operation "
+      << streamOptions;
+  return false;
+}
+
+std::optional<IndexNode*> getDependentIndexNode(
+    CollectNode const& collectNode) {
+  auto dependency = collectNode.getFirstDependency();
+  while (dependency != nullptr) {
+    LOG_RULE << "DEPENDENCY " << dependency->id() << " "
+             << dependency->getTypeString();
+    if (dependency->getType() == EN::INDEX) {
+      break;
+    } else if (dependency->getType() == EN::CALCULATION) {
+      auto calculation = EN::castTo<CalculationNode*>(dependency);
+      LOG_RULE << "CALCULATION NODE " << calculation->id();
+      dependency = dependency->getFirstDependency();
+    } else {
+      // maybe skip some nodes?
+      dependency = nullptr;
+    }
+  }
+
+  if (dependency == nullptr) {
+    LOG_RULE << "No suitable dependency found";
     return std::nullopt;
   }
```

Comment:
I think there is a slight flaw with this approach. Since we rely on the index that the `IndexNode` chose and not this optimization, we can use an index that is not good this rule, while there might be a better one.
e.g. If the collection has two indexes one sparse and another not, and the index node picked the sparse one, then this optimization fails.
I think a better approach would be to iterate thought all indexes on the collection and then choose the optimal one, but you decide if that is in this scope or not, I would at least leave a comment here.

Reply:
The index node could be chosen because of some other optimization rule that was applied earlier. If we would now exchange this index node here, we could destroy some other optimizations that were made before. This optimizer rule just tries to combine an existing collect and its dependent existing index node, which is kind of easy to do. But I added a comment at the top of the `useIndexForCollect` function to make this clear.